### PR TITLE
[core][state][telemetry] Add telemetry for top level state API usage from python. 

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: Ricky Xu <xuchen727@hotmail.com>

## Why are these changes needed?

Adding telemetry for state api usage. 
- This should only be collecting what functions (e.g. listing tasks,  listing actors) people are invoking. 
- We will NOT collect any other info of those resources other than the invocations existence. 
- This is to track the usage of state API, and help prioritize future improvements. 
- This should at most trigger the `number of added calls` to GCS if all tracked APIs are called (since it doesn't override with same values). 

TODO:
 -  We currently implement get based on list so we couldn't track from dashboard if get APIs are called. But this is probably fine for now. 

## Checks
Followed doc and 
```
{"usage_stats": {"ray_version": "3.0.0.dev0", "python_version": "3.8.13", "schema_version": "0.1", "source": "OSS", "session_id": "32ba512f-e0f4-4dfa-ab8a-df268dcc54da", "git_commit": "{{RAY_COMMIT_SHA}}", "os": "darwin", "collect_timestamp_ms": 1668276431427, "session_start_timestamp_ms": 1668276307449, "cloud_provider": null, "min_workers": null, "max_workers": null, "head_node_instance_type": null, "worker_node_instance_types": null, "total_num_cpus": 10, "total_num_gpus": null, "total_memory_gb": 33.778858948498964, "total_object_store_memory_gb": 2.0, "library_usages": ["dataset"], "total_success": 0, "total_failed": 119, "seq_number": 119, "extra_usage_tags": {"gcs_storage": "memory", "core_state_api": "list_tasks"}, "total_num_nodes": 1, "total_num_running_jobs": 1}, "success": false, "error": "HTTPConnectionPool(host='localhost', port=8000): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x10c825b80>: Failed to establish a new connection: [Errno 61] Connection refused'))"}
```

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
Should we just track it from the dashboard? Also I think this won't work because `record_extra_usage_tag ` requires the internal KV to be initialized. Also this doesn't work if the API is called outside the cluster (e.g., CLI used outside the head node)

**Comment by rickyyx** [comment]
@jjyao Does the `/tmp/ray/session_latest/usage_stats.json` only show the latest tag being reported?

**Comment by jjyao** [comment]
> @jjyao Does the /tmp/ray/session_latest/usage_stats.json only show the latest tag being reported?

Yes. It shows the latest report.

File: python/ray/_private/usage/usage_lib.py [hunk 1]
```diff
@@ -324,6 +328,24 @@ def record_library_usage(library_usage: str):
...
+        tag_key: TagKey defined
+        tag_value: string value
+
+    Example:
+        @record_usage(TagKey.MY_KEY, "MY_VAL")
+        def func(a, b):
+            ...
+
+    """
+    record_extra_usage_tag(tag_key, tag_value)
```
Comment (rkooo567) [inline python/ray/_private/usage/usage_lib.py:345]: Can you verify this is recorded? I think the process needs to have internal kv initialized

Comment (rickyyx) [reply python/ray/_private/usage/usage_lib.py:345]: How could I do that? I followed this basically https://docs.google.com/document/d/1XSWzx9ZatnlttsurCXzvV4vVsb2g1KxJ2T4zpf33bI8/edit not enough for testing?  cc @jjyao

Comment (rkooo567) [reply python/ray/_private/usage/usage_lib.py:345]: I think it is already initialized in the dashboard.

Comment (rkooo567) [reply python/ray/_private/usage/usage_lib.py:345]: is it coming up to the usage_stats.json file?

Comment (rickyyx) [reply python/ray/_private/usage/usage_lib.py:345]: Yep - see the description part.

**Review by scv119 — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Actually, does it work if you call 2 APIs? E.g., list_actors and list_nodes. Both of usage is recorded? From code, it seems like only the latest one will be recorded

**Comment by rickyyx** [comment]
> Actually, does it work if you call 2 APIs? E.g., list_actors and list_nodes. Both of usage is recorded? From code, it seems like only the latest one will be recorded

Oh good catch - i was wondering why it only reported latest one, thought it did recording.

**Review by rkooo567 — COMMENTED** [review]
Can you add a test here? test_usage_lib.py

File: python/ray/_private/usage/usage_lib.py [hunk 2]
```diff
@@ -338,6 +357,23 @@ def record_library_usage(library_usage: str):
         _put_library_usage(library_usage)
 
 
+@contextmanager
+def record_usage(tag_key: TagKey):
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply python/ray/_private/usage/usage_lib.py:361]: why moving this to dashboard?

**Review by pcmoritz — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Pleaes cherry pick it!

**Pull request merged**